[Lim Yixiang] iP

limyixiang opened the pull request:
### FriendlyBot

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

FriendlyBot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/limyixiang/ip?tab=readme-ov-file).
2. double-click it.
3. add your tasks.
4. let is manage your tasks for you 📖 

And it is **FREE**! 😉 

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon) 
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) {
        FriendlyBot friendlyBot = new FriendlyBot(TASK_LIST_FILE_PATH, TASK_LIST_FILE_PATH_WITHOUT_FILE);
        // Initialise variables
        Scanner reader = new Scanner(System.in);
        // Start of Chat Bot
        friendlyBot.ui.showWelcome();
        String response = "";
        boolean isExit = false;
        // Commands
        while (!isExit) {
            response = reader.nextLine();
            friendlyBot.ui.printHorizontalBar();
            Command cmd = Parser.parse(response);
            cmd.execute(friendlyBot.tasks, friendlyBot.ui, friendlyBot.storage);
            isExit = cmd.isExit();
            friendlyBot.ui.printHorizontalBar();
        }
        reader.close();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

limyixiang committed:
Renamed Chatbot to Friendly Bot.
Implemented skeletal version of greet and exit.

limyixiang committed:
Improved skeletal version:
Echoes commands entered by user and exits upon "bye"

limyixiang committed:
Added abiity for Friendly Bot to store tasks and return to user when requested.

limyixiang committed:
Added the ability to mark tasks as done.
Added the ability to mark tasks as undone.
Catches exceptions where given task index is out of range or when given task index is not an integer.

limyixiang committed:
Added support for tracking three types of tasks: todos, deadlines, events
Added new Classes for ToDo, Deadline and Event.
Added toString method for Task.
Added 3 new commands for todo deadline and event to create these task types.

limyixiang committed:
Added Automated Text UI Testing using the input/ouput redirection technique.

limyixiang committed:
FriendlyBot now correctly deals with incorrect inputs.

limyixiang committed:
Added support for deleting tasks in the list.
Changed the data structure used for storing tasks from using Array to ArrayList.

limyixiang committed:
Added JavaDocs for the following files.

limyixiang committed:
Save task in hard disk upon change

limyixiang committed:
Merge branch 'branch-Level-7'

limyixiang committed:
Change deadline and event to only accept yyyy-mm-dd as dates
Add command `date` to display events occurring on a specific date

limyixiang committed:
Merge branch 'branch-Level-8'

limyixiang committed:
Add more classes to make the code more Object-Oriented

limyixiang committed:
Merge branch 'branch-A-MoreOOP'

limyixiang committed:
Divide classes into packages

limyixiang committed:
Merge branch 'branch-A-Packages'

limyixiang committed:
Merge branch 'add-gradle-support' of https://github.com/limyixiang/ip

limyixiang committed:
Add JUnit Tests

limyixiang committed:
Edit mainClass in build.gradle

limyixiang committed:
Add JavaDoc for every file

limyixiang committed:
Tweak code to adhere to Coding Standard

limyixiang committed:
Add FindCommand

limyixiang committed:
Merge branch 'branch-A-JavaDoc'

limyixiang committed:
Merge branch 'branch-A-CodingStandard'

limyixiang committed:
Merge branch 'branch-Level-9'

limyixiang committed:
Update JavaDoc

limyixiang committed:
Update text-ui-test

limyixiang committed:
Add CheckStyle and correct styling bugs

limyixiang committed:
Merge branch 'branch-A-CheckStyle'

limyixiang committed:
Set up GUI using javafx

limyixiang committed:
Add support for commands within GUI

limyixiang committed:
Merge branch 'branch-Level-10'

limyixiang committed:
Fix style bug (missing javadoc)

SeeYangZhi commented on src/main/java/friendlybot/FriendlyBot.java:
`main` method is well-organized with proper Javadoc style comments, handling user input in a loop and executing commands. The `getResponse` method is a useful addition for integrating with a GUI, allowing the bot to process input and return responses

SeeYangZhi commented on src/main/java/friendlybot/Parser.java:
The `parser` class efficiently handles user input by mapping it to corresponding command objects. 

Comprehensive error handling for invalid inputs, and providing user-friendly messages. The use of multiple if-else statements could be refactored into a more scalable solution, such as a switch case.

limyixiang replied:
Thank you for your input! I will look into that 😄

SeeYangZhi commented on src/main/java/friendlybot/Storage.java:
`Storage` class includes robust error handling for file I/O operations, ensuring that the application can recover from errors by resetting to an empty task list. 

The `getTask` method effectively converts string representations of tasks into Task objects, supporting different task types.

SeeYangZhi commented on src/main/java/friendlybot/Ui.java:
`Ui` class provides methods for interacting with the user through console output. It includes methods for printing messages, welcome, and exit messages, as well as a horizontal bar for visual separation. The class is straightforward and serves its purpose well, though it could be expanded to handle more complex UI interactions if needed.

SeeYangZhi commented on src/main/java/friendlybot/command/Command.java:
Each command class encapsulates a specific action, adhering to the command pattern. This design makes it easy to extend the application with new commands. The execute methods are well-documented, explaining the interactions with TaskList, Ui, and Storage. The BadCommand class provides a graceful way to handle invalid inputs.

SeeYangZhi commented on src/main/java/friendlybot/task/Task.java:
The task classes represent different types of tasks, each with specific attributes. Good use of inheritance which allows for code reuse and a clear hierarchy. 

`toString` methods provide a consistent format for task representation, which is important for both user display and file storage.

SeeYangZhi reviewed (commented):
LGTM!

yuanchengg commented on src/main/java/friendlybot/Storage.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return An ArrayList of Tasks in the specified file path.
+     */
+    public ArrayList<Task> load() {
```
perhaps the naming for this method could be slightly clearer? and similar to your other method in this class e.g. loadFiles()

yuanchengg commented on src/main/java/friendlybot/command/MarkCommand.java:
```diff
@@ -0,0 +1,58 @@
+package friendlybot.command;
```
love how you use different command classes for different commands, great and easy way to update commands in the future 😄

yuanchengg commented on src/main/java/friendlybot/task/TaskList.java:
```diff
@@ -0,0 +1,130 @@
+     *
+     * @return A String representation of the list of tasks in TaskList.
+     */
+    public String formatTasksToSave() {
```
should this be handled in storage? could i know more on why you decide to place this method here? thanks!

yuanchengg reviewed (commented):
LGTM. Love your use of OOP in this code and how well you split them into classes, especially for commands.

limyixiang committed:
Add Assertions to code

Assertions are used to define assumptions about the program state so
that the runtime can verify them. An assertion failure indicates a
possible bug in the code as the code has resulted in a program state
that violates an assumption about how the code should behave.

Let's use the `assert` feature to document important assumptions that
should hold at various points in the code.

limyixiang committed:
Improve code quality

In any non-trivial software project, code needs to be read, understood,
and modified by other developers later on. Code quality is essential
for other developers, or even your future self.

We can enhance code quality by:
- Making the happy path prominent
- Structuring code logically
- Avoiding magic numbers
- Avoiding complicated expressions
- Avoiding deep nesting
- Avoiding long methods

limyixiang committed:
Add Streams

Streams allow developers to perform complex operations on collections of
data in a concise and elegant way, leading to more readable and
maintainable code.

Some methods in TaskList.java have been modified to support Stream
methods.

limyixiang committed:
Merge pull request #1 from limyixiang/branch-A-Assertions

Add Assertions to code

limyixiang committed:
Merge branch 'master' into branch-A-CodeQuality

limyixiang committed:
Merge pull request #2 from limyixiang/branch-A-CodeQuality

Improve code quality

limyixiang committed:
Merge branch 'master' into branch-A-Streams

limyixiang committed:
Merge pull request #3 from limyixiang/branch-A-Streams

Add Streams

limyixiang committed:
Merge branch 'master' of https://github.com/limyixiang/ip

limyixiang committed:
Add Continuous Integration

limyixiang committed:
Merge branch 'branch-A-CI'

limyixiang committed:
Update MarkCommand.java to fix missing print statement

limyixiang committed:
Update runtest.bat and runtest.sh

limyixiang committed:
Update runtest.bat and runtest.sh

limyixiang committed:
Merge branch 'master' of https://github.com/limyixiang/ip

limyixiang committed:
Add program exit on "bye" command

limyixiang committed:
Add help command

Help command allows the user to view the list of all existing
commands that FriendlyBot provides. For more information on each
command, the user can send `help <command>` to view the command's
usage information.

limyixiang committed:
Merge branch 'branch-BCD-Extension'

limyixiang committed:
Improve GUI to look more polished

limyixiang committed:
Merge branch 'branch-A-BetterGui'

limyixiang committed:
Add personality to FriendlyBot's replies

limyixiang committed:
Merge branch 'branch-A-Personality'

limyixiang committed:
Add more error handling

FriendlyBot now ignores leading and trailing whitespaces in
commands, and enforces that command format has to be met with
exactly one "/by" parameter in `deadline` command, and exactly one
"/from" and "/to" parameter in `event` command.

limyixiang committed:
Merge branch 'branch-A-MoreErrorHandling'

limyixiang committed:
Update EXPECTED.TXT's goodbye message

limyixiang committed:
Add more test cases

Some test cases were generated by GitHub CoPilot.

limyixiang committed:
Merge branch 'branch-A-MoreTesting'

limyixiang committed:
Improve code quality

limyixiang committed:
Add README.md

limyixiang committed:
Update return String of DateCommand

limyixiang committed:
Merge branch 'branch-A-UserGuide'

limyixiang committed:
Update README.md
